File: algorithm-and-data-structure-for-programing-contest/section2/alds_1_1d.cpp
#include <iostream>
#include <algorithm>
using namespace std;
static const int MAX = 200000;

/*
[sample input]
6
5
4
1
3
4
3

[sample output]
3
*/
int main(){
    int R[MAX], n;

    // 入力
    cin >> n;
    for(int i=0; i<n; i++){
        cin >> R[i];
    }

    // 実装
    int maxv = -2000000000;  // 十分小さい値
    int minv = R[0];

    for(int i=1; i<n; i++){
        maxv = max(maxv, R[i]-minv);
        minv = min(minv, R[i]);
    }

    cout << maxv << endl;

    return 0;
}

File: algorithm-and-data-structure-for-programing-contest/section2/alds1_1_d.py

if __name__ == "__main__":
    n = int(input())
    r0 = int(input())
    r1 = int(input())
    minv = r0 if r0 <= r1 else r1
    maxv = r1 - r0
    for i in range(2, n):
        r = int(input())
        maxv = max(maxv, (r-minv))
        minv = min(minv, r)
    print(maxv)
